flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/GroupAggProcessFunctionWithUpdateInterval.scala

```diff
@@ -0,0 +1,204 @@
+  protected var state: ValueState[Row] = _
+
+  // counts the number of added and retracted input records
+  protected var cntState: ValueState[JLong] = _
+
+  // stores the input for group keys
+  private var inputState: ValueState[Row] = _
+
+  // stores the last emit row
+  private var emitState: ValueState[Row] = _
+
+  // stores the emit time
+  private var emitTimerState: ValueState[JLong] = _
+
+
+  override def open(config: Configuration) {
+    LOG.debug(s"Compiling AggregateHelper: ${genAggregations.name}\n\n" +
+      s"Code:\n${genAggregations.code}")
+    val clazz = compile(
+      getRuntimeContext.getUserCodeClassLoader,
+      genAggregations.name,
+      genAggregations.code)
+    LOG.debug("Instantiating AggregateHelper.")
+    function = clazz.newInstance()
+
+    newRow = new CRow(function.createOutputRow, true)
+    prevRow = new CRow(function.createOutputRow, false)
+    typeSerializer = outputRowType.createSerializer(new ExecutionConfig())
+
+    state = getRuntimeContext.getState(
+      new ValueStateDescriptor[Row]("GroupAggregateState", aggregationStateType))
+    cntState = getRuntimeContext.getState(
+      new ValueStateDescriptor[JLong]("GroupAggregateInputCounter", Types.LONG))
+    inputState = getRuntimeContext.getState(
+      new ValueStateDescriptor[Row]("GroupAggregatePreEmitState", classOf[Row]))
+    emitState = getRuntimeContext.getState(
+      new ValueStateDescriptor[Row]("GroupAggregateEmitState", outputRowType))
+    emitTimerState = getRuntimeContext.getState(
+      new ValueStateDescriptor[JLong]("emitTimeState", Types.LONG))
+
+    initCleanupTimeState("GroupAggregateWithUpdateIntervalCleanupTime")
+  }
+
+  override def processElement(
+      inputC: CRow,
+      ctx: ProcessFunction[CRow, CRow]#Context,
+      out: Collector[CRow]): Unit = {
+
+    val currentTime = ctx.timerService().currentProcessingTime()
+    // register state-cleanup timer
+    registerProcessingCleanupTimer(ctx, currentTime)
+
+    val input = inputC.row
+
+    // get accumulators and input counter
+    var accumulators = state.value
+    var inputCnt = cntState.value
+
+    if (null == accumulators) {
+      accumulators = function.createAccumulators()
+      inputState.update(input)
+      inputCnt = 0L
+    }
+
+    // update aggregate result and set to the newRow
+    if (inputC.change) {
+      inputCnt += 1
+      // accumulate input
+      function.accumulate(accumulators, input)
+    } else {
+      inputCnt -= 1
+      // retract input
+      function.retract(accumulators, input)
+    }
+
+    state.update(accumulators)
+    cntState.update(inputCnt)
+
+    var triggerTimer = emitTimerState.value
+
+    if (null == triggerTimer) {
+      triggerTimer = 0L
+    }
+
+    if (currentTime >= triggerTimer) {
+
+      val newTimer = currentTime + queryConfig.getUnboundedAggregateUpdateInterval
+
+      emitTimerState.update(newTimer)
+
+      ctx.timerService().registerProcessingTimeTimer(newTimer)
```

Comment:
Yes, that's why I `registerProcessingTimeTimer` at the first-time. At current time, we only supported proc-time non-windowed aggregations. So, we can using `registerProcessingTimeTimer` for proc-time. When we supported event-time non-windowed aggregations we can using `registerEventTimeTimer`.  Because event-time non-windowed aggregations need an additional Implement. And I think we do not need `updateIntervalTimeMode` If we have different implementations for proc-time and event-time. 
i.e., `updateInterval <= 0 ` is immediate update. `updateInterval > 0 ` is Interval update.
What do you think?